### PR TITLE
bpo-40784: Fix sqlite3 deterministic test

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue40784

### CODE REVIEW

File: Lib/sqlite3/test/userfunctions.py
Comment (vstinner) [inline Lib/sqlite3/test/userfunctions.py]: I suggest to not match the whole error message, only check if it contains "deterministic", something like:

```suggestion
        # non-deterministic functions prohibited in partial index WHERE clauses
        self.assertIn("non-deterministic functions prohibited", str(cm.exception))
```

Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py]: I agree. I see from the CI that earlier versions of SQLite produces slightly different messages. I was thinking something like `self.assertOk(str(cm.exception).contains("prohibited"))`, but `self.assertIn()` is of course much more neat!

Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py]: Also, I see from the CI that the SQLite version on the Ubuntu machine in the Azure pipelines test actually disallows _any_ function in partial index clauses, so we need another way of asserting this. I tried with `CREATE TABLE test(t TEXT CHECK (nondeterministic())`, but I was unable to trigger an SQLite "exception". I'll do some more experiments in a few hours.

Comment (vstinner) [reply Lib/sqlite3/test/userfunctions.py]: Well, if tests still fail with self.assertIn("non-deterministic functions prohibited", str(cm.exception)), I suggest to even remove the test on the error message. Expecting an exception is already good enough.

**Comment by vstinner** [comment]
Maybe if would be safer to run tests on buildbots to check more SQLite versions, before merging this PR.

File: Lib/sqlite3/test/userfunctions.py
Comment (vstinner) [inline Lib/sqlite3/test/userfunctions.py:4]: What changed on this line? This diff looks suspicious :-)

File: Lib/sqlite3/test/userfunctions.py
Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py]: Sorry, for being unclear. The Azure pipeline test failed both at the deterministic and the non-deterministic test, but for different reasons:
- The deterministic test failed because for this specific SQLite version, even deterministic functions were disallowed for the given SQL expression.
- The non-deterministic test failed because the expected error message was slightly different.

File: Lib/sqlite3/test/userfunctions.py
Comment (encukou) [reply Lib/sqlite3/test/userfunctions.py:4]: It changes the `ä` from latin-1 to utf-8. So the change is good :)
(I checked with `git show` locally.)

Comment (hroncok) [reply Lib/sqlite3/test/userfunctions.py:4]:
```patch
-# Copyright (C) 2005-2007 Gerhard Häring <gh@ghaering.de>
+# Copyright (C) 2005-2007 Gerhard HÃ¤ring <gh@ghaering.de>
```

Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py:4]: It looks suspicious indeed :) Non-intentional! Will fix.

Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py:4]: Aha, it was GitHub! f7f9ef1a2423fe4e31e0f881fdaaf0816baf4d35

Comment (vstinner) [reply Lib/sqlite3/test/userfunctions.py:4]: Can you try to revert it?

Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py:4]: Sure, will do

Comment (vstinner) [reply Lib/sqlite3/test/userfunctions.py:4]: > It changes the ä from latin-1 to utf-8. So the change is good :)

Except that the file is encoded to latin-1 and starts with "#-*- coding: iso-8859-1 -*-" header.

Comment (vstinner) [reply Lib/sqlite3/test/userfunctions.py:4]: If someone wants to convert this file to UTF-8, #-- coding: iso-8859-1 --" header must be removed.

Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py:4]: That calls for a separate PR, @vstinner ?

Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py:4]: BTW, reverted in 46642792956cbba07903ae66ae26e21a4462d16f

**Review by vstinner — APPROVED** [review]
LGTM if CI pass. As I wrote, it sounds worth it to spawn buildbots tests on this PR.

@erlend-aasland: Please stop touching your PR :-) So we can run more tests (buildbots are likely to use more various SQLite versions).

**Comment by vstinner** [comment]
Oh, it may also help if someone could test manually with SQLite 3.32.

File: Lib/sqlite3/test/userfunctions.py
Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py]: Another possibility would be to require a newer SQLite version. After all, it's been a while since 3.7.3 was released...

**Comment by erlend-aasland** [comment]
> Oh, it may also help if someone could test manually with SQLite 3.32.

Already did that, and it passes! 3.32.1 FWIW

File: Lib/sqlite3/test/userfunctions.py
Comment (vstinner) [reply Lib/sqlite3/test/userfunctions.py:4]: > That calls for a separate PR, @vstinner ?

Technically, the file is properly encoded and valid. I don't see any **need** to change it. If it's not broken, don't fix it :-)

**Comment by vstinner** [comment]
> Already did that, and it passes! 3.32.1 FWIW

Cool. Thanks!

File: Lib/sqlite3/test/userfunctions.py
Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py]: This still fails. I've fixed `CheckFuncNonDeterministic` locally but not pushed it yet. FYI:
```
-        self.assertIn("non-deterministic functions prohibited", str(cm.exception))
+        self.assertIn("prohibited", str(cm.exception))
```

However, `CheckFuncDeterministic` still fails, as I described in an earlier comment. I'll see if I can find a workaround. I won't push anything until the `CheckFuncDeterministic` issue is resolved. No need to run the CI again when we know it will fail.

**Comment by vstinner** [comment]
> This still fails. I've fixed CheckFuncNonDeterministic locally but not pushed it yet.

Yeah, the Ubuntu job of Azure Pipelines fails. It uses "sqlite3.sqlite_version: 3.11.0" according to its test.pythoninfo. Test failure:

```
FAIL: CheckFuncNonDeterministic (sqlite3.test.userfunctions.FunctionTests)
Non-deterministic functions cannot be used with partial indices
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/Lib/sqlite3/test/userfunctions.py", line 288, in CheckFuncNonDeterministic
    self.assertIn("non-deterministic functions prohibited", str(cm.exception))
AssertionError: 'non-deterministic functions prohibited' not found in 'functions prohibited in partial index WHERE clauses'
```

Honestly, I don't think that we should test the error message. Please simply **remove** the test the error message. IMHO excepting an exception is enough.

**Comment by erlend-aasland** [comment]
> Yeah, the Ubuntu job of Azure Pipelines fails. It uses "sqlite3.sqlite_version: 3.11.0" according to its test.pythoninfo.

Right. Deterministic functions were allowed in partial indices in sqlite v3.15.0 (2016-10-14). I'll find another way of testing this, then.

> Honestly, I don't think that we should test the error message. Please simply **remove** the test the error message. IMHO excepting an exception is enough.

Agreed. Error messages are likely to change between versions.

**Comment by erlend-aasland** [comment]
@vstinner I suggest just keeping the old tests for SQLite pre v3.15.0. From the sqlite release history:
```
2014-02-03 (3.8.3) […] Added SQLITE_DETERMINISTIC […] providing applications with the ability to create new functions that can be factored out of inner loops when they have constant arguments.
```

Seven versions later:
```
2016-10-14 (3.15.0) […] Allow deterministic SQL functions in the WHERE clause of a partial index.
```

I'll push the change right away.

File: Lib/sqlite3/test/userfunctions.py
Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py]: Resolved in b969ae5a60d984dde146e01e903b9d0bacf909a7 
Tested with sqlite 3.32.1, 3.15.0, 3.14.2 and 3.11.0.

**Comment by encukou** [comment]
Thank you!
I'll start the buildbots so Victor can continue the review when he gets to the computer.

**Comment by erlend-aasland** [comment]
Thank you, @encukou!

**Comment by encukou** [comment]
Some failures might not be related to this change.

**Comment by vstinner** [comment]
AMD64 Fedora Stable Clang Installed PR, PPC64LE Fedora Stable Clang Installed PR and s390x Fedora Clang Installed PR and x86 Gentoo Installed with X PR failed because of https://bugs.python.org/issue38488 regression (setuptools/chmod) which is unrelated.

AMD64 FreeBSD Non-Debug PR failures (test_asyncio, test_venv) are unrelated.

AMD64 Windows8.1 Refleaks PR failure (test_socket) is unrelated.

File: Misc/NEWS.d/next/Library/2020-05-27-11-17-02.bpo-40784.hYEBjs.rst
Comment (berkerpeksag) [inline Misc/NEWS.d/next/Library/2020-05-27-11-17-02.bpo-40784.hYEBjs.rst]: A NEWS entry is only needed for user-facing bugs and features. This PR only fixes a test. I've just set 'skip news' label so you can safely delete it.

**Review by berkerpeksag — CHANGES_REQUESTED** [review]

File: Misc/NEWS.d/next/Library/2020-05-27-11-17-02.bpo-40784.hYEBjs.rst
Comment (vstinner) [reply Misc/NEWS.d/next/Library/2020-05-27-11-17-02.bpo-40784.hYEBjs.rst]: It's fine to document fixed tests. But it may be good to move this NEWS entry into the Tests category, and maybe explain that it fix the test for SQLite 3.32.

I'm also fine with removing the NEWS entry ;-)

Comment (erlend-aasland) [reply Misc/NEWS.d/next/Library/2020-05-27-11-17-02.bpo-40784.hYEBjs.rst]: Thanks! Noted for future reference :)

**Comment by erlend-aasland** [comment]
I have made the requested changes; please review again

**Review by berkerpeksag — APPROVED** [review]
LGTM! Thanks for fixing this. I just made a small whitespace change and will merge once tests are passed (unless Victor or Petr have something else to say)

File: Lib/sqlite3/test/userfunctions.py
Comment (vstinner) [inline Lib/sqlite3/test/userfunctions.py:4]: It seems like the UTF-8 issue was reintroduced.

**Review by vstinner — CHANGES_REQUESTED** [review]
Please fix the copyright line. Either revert the change, or remove the coding header (1st line) (and ensure that the new file is properly encoded to UTF-8).

File: Lib/sqlite3/test/userfunctions.py
Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py:4]: That's very strange; I can't find that change in my repo. If it's ok with you, I'll re-encode the file with UTF-8.

Comment (erlend-aasland) [reply Lib/sqlite3/test/userfunctions.py:4]: Aha, it was @berkerpeksag's commit f5706584c0e785e3a3b91645dec8329bc9b257f6 (apparently edited with GitHub; same problem as yesterday).

**Comment by erlend-aasland** [comment]
I have made the requested changes; please review again

**Comment by erlend-aasland** [comment]
Thanks for the review, @berkerpeksag & @vstinner !

**Review by berkerpeksag — APPROVED** [review]

**Comment by berkerpeksag** [comment]
Thanks for the PR!

**Comment by vstinner** [comment]
Thanks for fixing the encoding/copyright line ;-)

**Pull request merged**
